chia-protocol/src/program.rs

```diff
@@ -59,6 +50,270 @@ impl Program {
+fn clvm_convert(a: &mut Allocator, o: &PyAny) -> PyResult<NodePtr> {
+    // None
+    if o.is_none() {
+        Ok(a.null())
+    // Program itself
+    } else if let Ok(prg) = o.extract::<Program>() {
+        Ok(node_from_bytes_backrefs(a, prg.0.as_slice())?)
+    // bytes
+    } else if let Ok(buffer) = o.extract::<&[u8]>() {
+        a.new_atom(buffer)
+            .map_err(|e| PyMemoryError::new_err(e.to_string()))
+    // str
+    } else if let Ok(text) = o.extract::<String>() {
+        a.new_atom(text.as_bytes())
+            .map_err(|e| PyMemoryError::new_err(e.to_string()))
+    // int
+    } else if let Ok(val) = o.extract::<clvmr::number::Number>() {
+        a.new_number(val)
+            .map_err(|e| PyMemoryError::new_err(e.to_string()))
+    // Tuple (SExp-like)
+    } else if let Ok(pair) = o.downcast::<PyTuple>() {
+        if pair.len() != 2 {
+            Err(PyValueError::new_err(format!(
+                "can't cast tuple of size {}",
+                pair.len()
+            )))
+        } else {
+            let left = clvm_convert(a, pair.get_item(0)?)?;
+            let right = clvm_convert(a, pair.get_item(1)?)?;
+            a.new_pair(left, right)
+                .map_err(|e| PyMemoryError::new_err(e.to_string()))
+        }
+    // List
+    } else if let Ok(list) = o.downcast::<PyList>() {
+        let mut rev = Vec::<&PyAny>::new();
+        for py_item in list.iter() {
+            rev.push(py_item);
+        }
+        let mut ret = a.null();
+        for py_item in rev.into_iter().rev() {
+            let item = clvm_convert(a, py_item)?;
+            ret = a
+                .new_pair(item, ret)
+                .map_err(|e| PyMemoryError::new_err(e.to_string()))?;
+        }
+        Ok(ret)
+    // SExp (such as clvm.SExp)
+    } else if let (Ok(atom), Ok(pair)) = (o.getattr("atom"), o.getattr("pair")) {
+        if atom.is_none() {
+            if pair.is_none() {
+                Err(PyTypeError::new_err(format!("invalid SExp item {o}")))
+            } else {
+                let pair = pair.downcast::<PyTuple>()?;
+                let left = clvm_convert(a, &pair[0])?;
+                let right = clvm_convert(a, &pair[1])?;
+                a.new_pair(left, right)
+                    .map_err(|e| PyMemoryError::new_err(e.to_string()))
+            }
+        } else {
+            a.new_atom(atom.extract::<&[u8]>()?)
+                .map_err(|e| PyMemoryError::new_err(e.to_string()))
+        }
+    // anything convertible to bytes
+    } else if let Ok(fun) = o.getattr("__bytes__") {
+        let bytes = fun.call0()?;
+        let buffer = bytes.extract::<&[u8]>()?;
+        a.new_atom(buffer)
+            .map_err(|e| PyMemoryError::new_err(e.to_string()))
+    } else {
+        Err(PyTypeError::new_err(format!(
+            "unknown parameter to run_with_cost() {o}"
+        )))
+    }
+}
+
+#[cfg(feature = "py-bindings")]
+fn to_program(py: Python<'_>, node: LazyNode) -> PyResult<&PyAny> {
```

Comment:
This is not great, but in order to be compatible with the current Python version of `SerializedProgram` we need to be able to convert to and from the `Program` type, which is based on `SExp`